### PR TITLE
Add MIMIC-IV FHIR (NDJSON) dataset, MPF clinical prediction task, and EHRMambaCEHR example pipeline

### PR DESCRIPTION
**Pull request by evanfebrianto**
## Contributor

- **Name:** Evan Febrianto
- **NetID / email:** evanf3 / evanf3@illinois.edu

## Contribution type

**Full pipeline (dataset + task + model)** for reproducing **EHRMamba**-style training on **tokenized FHIR** timelines: CEHR-aligned features, optional **MPF** boundary tokens, and a **Mamba** sequence backbone.

## Relation to upstream

This PR **does not re-implement the core EHRMamba / Mamba stack from scratch**. It builds on **#809** (`EHRMamba`, `MambaBlock` in `pyhealth/models/ehrmamba.py`). New work is the **FHIR ingest**, **MPF clinical task**, **CEHR embedding variant + `EHRMambaCEHR`**, tests, docs, and the example script.

## Paper (reproducibility target)

- Fallahpour et al., *EHRMAMBA: Towards Generalizable and Scalable Foundation Models for Electronic Health Records* (ML4H 2024).
- **arXiv:** https://arxiv.org/abs/2405.14567

## Summary

End-to-end path for **MIMIC-IV on FHIR (NDJSON)**:

1. **`MIMIC4FHIRDataset`** — Grouped FHIR R4 resources → time-sorted CEHR-style token sequences; defaults in `pyhealth/datasets/configs/mimic4_fhir.yaml`.
2. **`MPFClinicalPredictionTask`** — Binary clinical prediction; optional MPF specials (`<mor>` / `<reg>`) vs generic (`<cls>` / `<reg>`).
3. **`EHRMambaCEHR`** — `BaseModel` combining **CEHR-style embeddings** (concept, token type, time, age, visit order, segment), existing **Mamba** blocks, and masked pooling for classification.

Also includes **synthetic NDJSON** for tests, **API RST** + index updates, and **`examples/mimic4fhir_mpf_ehrmamba.py`** (`--quick-test`, `--ablation`, or `--fhir-root` / `MIMIC4_FHIR_ROOT`).

## Reproduction results (credentialled MIMIC-IV on FHIR)

Partial numerical reproduction of the mortality task from the paper.

| Metric  | This PR (5 K patients, 2 Mamba blocks, mortality only) | Paper (~180 K patients, 32 blocks, 6 MPF tasks) |
|---------|:------------------------------------------------------:|:-----------------------------------------------:|
| ROC-AUC | **0.9129**                                             | 0.976                                           |
| PR-AUC  | **0.5711**                                             | 0.873                                           |
| Loss    | 0.2009                                                 | —                                               |

**Run configuration.** 5 000 MIMIC-IV-on-FHIR patients (PhysioNet credentialled), single-task MPF mortality fine-tune, 100 epochs, Adam lr=1e-3, `max_len=512`, `hidden_dim=128`, 2 Mamba blocks, H200 spot instance.

**Differences from the paper.** No MLM/NTP pre-training phase, 2 blocks vs 32, one clinical task vs 6 joint MPF tasks, ~36× smaller cohort. The remaining gap (0.913 → 0.976) is consistent with those scale/pre-training differences; closing it is pure scale-up (no code changes).

**Training loss.** Mean train loss decreases monotonically from ~0.25 to ~0.015 over 100 epochs with no divergence — evidence the pipeline learns the right signal rather than memorizing.

## Files to review

| Topic | Path(s) |
|-------|---------|
| FHIR ingest, sequences, dataset | `pyhealth/datasets/mimic4_fhir.py`, `pyhealth/datasets/fhir_utils.py` |
| YAML defaults | `pyhealth/datasets/configs/mimic4_fhir.yaml` |
| CEHR processor (tokenization + vocab + labels) | `pyhealth/processors/cehr_processor.py` |
| Task | `pyhealth/tasks/mpf_clinical_prediction.py` |
| CEHR embeddings + model | `pyhealth/models/cehr_embeddings.py`, `pyhealth/models/ehrmamba_cehr.py` |
| Mamba blocks (upstream) | `pyhealth/models/ehrmamba.py` |
| Tests | `tests/core/test_mimic4_fhir_dataset.py`, `tests/core/test_mpf_task.py`, `tests/core/test_ehrmamba_cehr.py`, `tests/core/test_mimic4_fhir_ndjson_fixtures.py` |
| Docs + TOC | `docs/api/datasets/pyhealth.datasets.MIMIC4FHIRDataset.rst`, `docs/api/tasks/pyhealth.tasks.mpf_clinical_prediction.rst`, `docs/api/models/pyhealth.models.EHRMambaCEHR.rst`; `docs/api/datasets.rst`, `tasks.rst`, `models.rst` |
| Example | `examples/mimic4fhir_mpf_ehrmamba.py` |

## How to run

From the repo root:

```bash
# Recommended: env with editable pyhealth (see pyproject.toml features)
pixi run -e base python examples/mimic4fhir_mpf_ehrmamba.py --quick-test
pixi run -e base python examples/mimic4fhir_mpf_ehrmamba.py --quick-test --ablation

# Alternative: default pixi env + source tree on PYTHONPATH
PYTHONPATH=. pixi run python examples/mimic4fhir_mpf_ehrmamba.py --quick-test

# Credentialled MIMIC-IV on FHIR (local NDJSON tree) — reproduces the 0.913 AUROC run
export MIMIC4_FHIR_ROOT=/path/to/fhir
pixi run -e base python examples/mimic4fhir_mpf_ehrmamba.py \
  --fhir-root "$MIMIC4_FHIR_ROOT" \
  --max-patients 5000 --max-len 512 --hidden-dim 128 \
  --epochs 100 --lr 1e-3
```

## Ablation harness

`--ablation` sweeps a small `max_len × use_mpf × hidden_dim` grid on the synthetic NDJSON fixture (1 epoch each, seconds in CI). The grid is wired to isolate the two hypotheses the paper makes:

- **Architectural hypothesis** — scaling `max_len` tests linear-memory context extension. The 5 K-patient real-data run above verifies the same switch scales to 512 with no code changes.
- **MPF hypothesis** — toggling `use_mpf` swaps only the first boundary token (`<mor>` vs generic `<cls>`), so any downstream delta is attributable to MPF alone.

## Testing

- Unit tests use *synthetic / in-memory NDJSON* only (no checked-in PHI).
- `pixi run test` to run.

## Attribution

- **CEHR / Odyssey-style** embedding layout informed by the CEHR-BERT / Odyssey ecosystem; integration is PyHealth-native.
- **Mamba / EHRMamba** core blocks reused from existing PyHealth code (#809).

### CODE REVIEW

File: pyhealth/processors/cehr_processor.py [hunk 1]
```diff
@@ -0,0 +1,364 @@
+"""CEHR-style tokenization, vocabulary, and sequence building for FHIR timelines.
```
Comment (jhnwu3) [inline pyhealth/processors/cehr_processor.py:1]: Just to confirm, this takes raw table data -> tokenizer -> tensors?

If so, then, this belongs in either the pyhealth.tasks.mpf_clinical_predictiont hat you've defined here, or you can define a separate CEHR processor that creates a new schema type here like "cehr" or something of that sort, see this example here:

on tokenizing language + time, etc. into one type of tensor or object for downstream processing:
https://github.com/sunlabuiuc/PyHealth/blob/master/pyhealth/processors/tuple_time_text_processor.py

File: pyhealth/datasets/fhir_ingest.py [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+    Stream all matching NDJSON/NDJSON.GZ resources into six per-type Parquet tables.
+
+sorted_ndjson_files(root, glob_pattern)
+    List matching NDJSON files under root (deduplicated, sorted).
+
+filter_flat_tables_by_patient_ids(source_dir, out_dir, keep_ids)
+    Subset existing flattened tables to a specific patient cohort.
+
+synthetic_mpf_two_patient_ndjson_text()
+    In-memory two-patient fixture for tests and quick-test mode.
```
Comment (jhnwu3) [inline pyhealth/datasets/fhir_ingest.py]: we can definitely just move these into the test cases.

File: pyhealth/datasets/fhir_utils.py [hunk 3]
```diff
@@ -0,0 +1,500 @@
+"""FHIR NDJSON parsing, flattening, and Parquet table writing.
```
Comment (jhnwu3) [inline pyhealth/datasets/fhir_utils.py:1]: maybe, we can rename this file to fhir_utils.py here?  I like fhr_ingest.py, but thinking of the future when there's going to be probably more adjustments to this as we expand fhir support.

File: pyhealth/datasets/mimic4_fhir.py [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+                .select("patient_id")
+                .unique()
+                .sort("patient_id")
+                .collect(engine="streaming")["patient_id"]
+                .to_list()
+            )
+            logger.info("Found %d unique patient IDs", len(self._unique_patient_ids))
+        return self._unique_patient_ids
+
+    def set_task(
```
Comment (jhnwu3) [inline pyhealth/datasets/mimic4_fhir.py]: If it truly follows the rest of the PyHealth pipeline here, are all of these override functions still necessary? If so, why?

Comment (jhnwu3) [reply pyhealth/datasets/mimic4_fhir.py]: by that i mean the ones below

**Review by jhnwu3 — CHANGES_REQUESTED** [review]
Thanks for making it more digestible. 

I've added my comments here, let me know if you still have more questions about this. I think we're pretty close. I just think there's still some misconceptions around processors and tasks here, and how we should link up everything here together for when we do downstream training.

File: pyhealth/datasets/fhir_utils.py (hunk 3 above)
Comment (evanfebrianto) [reply pyhealth/datasets/fhir_utils.py:1]: renamed and adjusted

File: pyhealth/processors/cehr_processor.py (hunk 1 above)
Comment (evanfebrianto) [reply pyhealth/processors/cehr_processor.py:1]: you're right, this is better to be in the processor instead.

File: pyhealth/datasets/fhir_ingest.py (hunk 2 above)
Comment (evanfebrianto) [reply pyhealth/datasets/fhir_ingest.py]: moved to tests

File: pyhealth/datasets/mimic4_fhir.py (hunk 4 above)
Comment (evanfebrianto) [reply pyhealth/datasets/mimic4_fhir.py]: I've changed this as well
